### PR TITLE
Tighten phone number validation

### PR DESCRIPTION
**Pull request by JonF1**
## Context
User could add any 50 numbers without a validation being triggered

## Changes proposed in this pull request
* Follow GovUK Notify rules by validating the number of digits without the the special characters `+-()` and whitespace. 
* There should be only 8-15 digits

## Guidance to review

Visit `candidate/application/contact-information/phone-number/edit` and check that phone number follows new validation rules. 

Do we want to add these rules to the support UI? This is just for normal candidate application form.

## Link to Trello card

https://trello.com/c/wGxge2qQ/4226-tighten-phone-number-validation

## Things to check

- [x] If the code removes any existing feature flags, a data migration has also been added to delete the entry from the database
- [x] This code does not rely on migrations in the same Pull Request
- [x] If this code includes a migration adding or changing columns, it also backfills existing records for consistency
- [x] API release notes have been updated if necessary
- [x] Required environment variables have been updated [added to the Azure KeyVault](/docs/environment-variables.md#deploy-pipeline)

### CODE REVIEW

**Review by tomas-stefano — APPROVED** [review]

File: app/forms/candidate_interface/contact_details_form.rb [hunk 1]
```diff
@@ -35,6 +39,15 @@ def self.build_from_application(application_form)
...
     end
 
+    def sanitize_phone_number
+      self.raw_phone_number = phone_number
+      self.phone_number = phone_number.gsub(/[\s()\-+]/, '') unless phone_number.nil?
+    end
+
+    def reset_phone_number
+      self.phone_number = raw_phone_number
+    end
```
Comment (frankieroberto) [inline app/forms/candidate_interface/contact_details_form.rb:49]: This feels a bit unexpected to me. Could we move this logic to the `PhoneNumberValidator` class instead, and move the length constraints there too?

Something like this maybe?

```ruby
class ModelName
  validates :phone_number, phone_number: true
end

class PhoneNumberValidator < ActiveModel::EachValidator
  def validate_each(record, attribute, value)
    if value.blank?
      record.errors.add(attribute, :blank)
    elsif invalid_phone_number_format?(value)
      record.errors.add(attribute, :invalid)
    elsif not_enough_digits?
      record.errors.add(attribute, :too_short)
    elsif too_many_digits?
      record.errors.add(attribute, :too_long)
    end
  end
  
  private
  
  def invalid_phone_number_format?(value)
    value.match?(/[^ext()+. 0-9]/)
    value.match?(/[^-()+. 0-9]/)
  end
  
  def not_enough_digits?
    value.gsub(/[^0-9]/, '').length < 8
  end

  def too_many_digits?
    value.gsub(/[^0-9]/, '').length > 15
  end
   
end
```

That way we'd also avoid displaying two messages (one about length and one about format) for the same field – better to only show one at a time.

File: spec/forms/candidate_interface/contact_details_form_spec.rb [hunk 2]
```diff
@@ -112,6 +112,25 @@
...
+
+  describe '#reset_phone_number' do
+    it 'returns raw phone number from user' do
+      form_data = { phone_number: '123456789123456', raw_phone_number: '123456789123456-+()' }
+      phone_numbers = described_class.new(form_data)
+      phone_numbers.reset_phone_number
+
+      expect(phone_numbers.phone_number).to eq(phone_numbers.raw_phone_number)
+    end
+  end
```
Comment (frankieroberto) [inline spec/forms/candidate_interface/contact_details_form_spec.rb:132]: Move these tests to the `phone_number_validator_spec.rb`?  Feels wrong to be testing methods called via callbacks – they should probably be private?

**Review by frankieroberto — CHANGES_REQUESTED** [review]

File: app/forms/candidate_interface/contact_details_form.rb (hunk 1 above)
Comment (JonF1) [reply app/forms/candidate_interface/contact_details_form.rb]: Thanks Frankie, we fixed this in 698eaea6a86443db9e441f48f9acee85e770ee5b

File: spec/forms/candidate_interface/contact_details_form_spec.rb (hunk 2 above)
Comment (JonF1) [reply spec/forms/candidate_interface/contact_details_form_spec.rb]: We've re-written these to be unit tests for `phone_number_validator_spec.rb`

File: config/locales/candidate_interface/contact_details.yml [hunk 3]
```diff
@@ -47,7 +47,8 @@ en:
             phone_number:
               blank: Enter your phone number
               invalid: Enter a phone number, like 01632 960 001, 07700 900 982 or +44 0808 157 0192
-              too_long: Phone number must be %{count} characters or fewer
+              too_long: Phone number must be 15 characters or fewer
+              too_short: Phone number must be 8 characters or more
```
Comment (frankieroberto) [inline config/locales/candidate_interface/contact_details.yml:51]: Maybe we should make this slightly more specific, as "characters" makes less sense for phone numbers?

```suggestion
              too_long: Phone number must contain 15 numbers or fewer
              too_short: Phone number must contain 8 numbers or more
```

**Review by frankieroberto — APPROVED** [review]

**Pull request merged**